Include the OS release info in backup directory names

rpi-simonz opened the pull request:
Adding to issue #778 this is my attempt implementing a different naming scheme of the backup directories which includes the OS version.

Using
```
raspberrypi@debian12-rsync-backup-20240416-094106
```
instead of
```
raspberrypi-rsync-backup-20240416-094106
```

This way every time a new OS version is saved a new backup directory is created/used and thus a new full backup is created.

Possibly already existing backups with the old name (without OS release info) are simply ignored during smartRecycle and while checking the number of backups to keep.

I successfully tested backup and restore in different configurations.

rpi-simonz committed:
Include OS release info in backup directory names

rpi-simonz committed:
Remove superfluous hostname from message

rpi-simonz committed:
smartRecycle: Keep old backups without OS release info in its name untouched

rpi-simonz committed:
Restore: Revert wrong modification and fix HOSTNAME detection

rpi-simonz committed:
Adapt findNonpartitionBackupBootAndRootpartitionFiles()

rpi-simonz commented:
Regarding this ...

> Possibly already existing backups with the old name (without OS release info)
> are simply ignored during smartRecycle and while checking the number of backups to keep.

... I'll check the other possible option: including the old ones. 

Probably tomorrow. Stay tuned.

rpi-simonz committed:
Merge branch 'framps:master' into hostname_osr

framps commented:
Appreciate this PR. It's very useful to include the OS release in the backup bath name. 

Unfortunately I'm busy testing the next release 0.7.0 and therefore don't accept the PR as of now. Please be patient. I'll accept the PR when I finished my tests for next release and add this nice feature in next release.

rpi-simonz commented:
I'm still working on some implementation details of this PR. 
So it's a perfect fit that you are busy elsewhere for some time!   ;-)

rpi-simonz committed:
Recycle: (Re-)Include old backups without OS release info into strategy

rpi-simonz committed:
Fake mode in normal backup type: Add info about backup recycling(deletion)

rpi-simonz commented:
It's done. As far as I can tell...

But considering the current discussion in https://forum-raspberrypi.de regarding partitionbased backups and their handling during recycling it might be an idea to 
- add an indicator for those backups into the directory name as well; like _rsync-pb-backup_

rpi-simonz committed:
Mostly revert to version 1 and let old backups untouched.
A user message will follow soon.
Improve getOSRelease to make it more robust.

rpi-simonz committed:
Fake mode in normal backup type: Add info about backup recycling(deletion), fixed

rpi-simonz committed:
Remove doubled local definition

rpi-simonz committed:
First version of reportOldBacks()

rpi-simonz committed:
Change check for file os-release

rpi-simonz committed:
Cleanup and improve reportOldBackups. Still not finished...

rpi-simonz committed:
Cleanup and change getOSRelease a bit

rpi-simonz committed:
Fix call location for reportOldBackups

rpi-simonz committed:
Improve message text in reportOldBackups

rpi-simonz committed:
Only show message if there are old-type backups existing

rpi-simonz committed:
New messages 400 and 401

rpi-simonz committed:
reportOldBackups now with writeToConsole messages, 402-406

rpi-simonz commented:
So, since in the meantime the strategy had been changed (again) it's now like this:

Old-type backups (without OS-release info in its name) are old backups. 
They are not touched/deleted/recycled automatically. 

But there is a - hopefully - clear message for the user about them: They can be deleted manually.

I used and created new MSG_ messages for writeToConsole as good as I could. 
Probably the numbering has to be adapted - I didn't know your preferences there.

framps commented:
> But there is a - hopefully - clear message for the user about them: They can be deleted manually.

I frankly thought to document this but usually docs are not read - so it's a very good idea to write a message to inform about this change.

> Probably the numbering has to be adapted - I didn't know your preferences there.

No worries - I also have added some new messages. Message numbers can be updated very easy :wink:

rpi-simonz committed:
Fix typo

rpi-simonz committed:
Fix typos in and enhance messages used in reportOldBackups

rpi-simonz committed:
Fix some indentation inconsistencies

rpi-simonz commented:
I enhanced the message about the directory name change by referring to a raspiBackup version number.
Of course that number needs to be checked again on implementation.

rpi-simonz committed:
Rename and reorder some of the new messages

rpi-simonz commented:
Commit 778ffc6 (Fix some indentation inconsistencies)  above isn't exactly related to this topic here. But it slipped in...

framps commented on raspiBackup.sh:
```diff
@@ -5606,7 +5664,7 @@ function backupRsync() { # partition number (for partition based backup)
 
 		bootPartitionBackup
-		lastBackupDir=$(find "$BACKUPTARGET_ROOT" -maxdepth 1 -type d -name "*-$BACKUPTYPE-*" ! -name $BACKUPFILE 2>>/dev/null | sort | tail -n 1)
+		lastBackupDir=$(find "$BACKUPTARGET_ROOT" -maxdepth 1 -type d -name "${HOSTNAME_OSR}*-$BACKUPTYPE-*" ! -name $BACKUPFILE 2>>/dev/null | sort | tail -n 1)
```
Shouldn't this be `${HOSTNAME_OSR}` without trailing `*`?

rpi-simonz replied:
Indeed! Fortunately the '*' doesn't hurt.  ;-) 
I will remove it.

framps commented on raspiBackup.sh:
```diff
@@ -6189,6 +6267,60 @@ function applyBackupStrategy() {
 	logExit
 }
 
+function reportOldBackups() {
```
This works perfect. But it will become annoying if it's reported forever. Please check how the RESTORE_REMINDER reminds for a limited times.

rpi-simonz replied:
It should be annoying, IMHO.
The (getting) old backups consume disk space and should be cleaned up soon.
But of course I'll check RESTORE_REMINDER.

framps replied:
> It should be annoying, IMHO.

:laughing: I don't agree. It's a very good idea to remind raspiBackup users. I'll going to document this also an my website but a lot of people don't read docs :cry: But it's not a good idea to bother them forever as long as they keep old fashioned backups. Keep in mind users who use the intelligent backup strategy and keep old backups for months and years. 

I'm sure issues will be created to stop these warnings.

framps replied:
I don't think it's required to have the reminder to be configurable in a config file as it's done for the reminder. Just use some constant (3,7,42 :wink: )

rpi-simonz replied:
> Keep in mind users who use the intelligent backup strategy and keep old backups for months and years.

Ups, yes. Longterm usage... 
I'll think about a not so annoying solution.

rpi-simonz replied:
> Just use some constant (3,7,42 😉 )

Questions I'm currently scratching my head about:

Should such a counter count for

- the general executions of raspiBackup on that host (saving it in the backup dir HOSTNAME)
- or individually per backup type (rsync, dd, etc.)
- or per backup strategy (normal / smart)?
- and even more possibilities...

I'll need to think a bit longer about that.

framps replied:
Good catch. But I wouldn't spend to much effort in this. Keep it simple.

Why don't you just check whether old files exist and count the number of events independent of the hostname? 

Purpose of this warning is to notify of this name change and old named directories are not recycled any more. Then the user knows about this and should check all existing backups. The warning text then should become more general and not host specific.

framps replied:
Great. Thank you very much :+1: . I'm still testing the new release code where I plan to integrate your PR. When I'm done I'll review your code carefully :wink:

framps commented on raspiBackup.sh:
```diff
@@ -7190,7 +7323,7 @@ function doitBackup() {
 
 	if (( $SMART_RECYCLE_DRYRUN && $SMART_RECYCLE )); then # just apply backup strategy to test smart recycle
-		writeToConsole $MSG_LEVEL_MINIMAL $MSG_APPLYING_BACKUP_STRATEGY_ONLY "$BACKUPPATH/$(hostname)"
+		writeToConsole $MSG_LEVEL_MINIMAL $MSG_APPLYING_BACKUP_STRATEGY_ONLY "$BACKUPPATH"
```
Why did you remove ${hostname} ?

rpi-simonz replied:
Because the hostname is already visible in the message. In the directory name.

rpi-simonz committed:
Delete overlooked asterisk

rpi-simonz committed:
First adaption of raspiBackup7412Test.sh - Untested yet

rpi-simonz committed:
Add TODO hints to raspiBackup7412Test.sh

rpi-simonz committed:
Fix code position of function

rpi-simonz commented:
At least raspiBackup7412Test.sh is running now - without complaints. 
But I'm not sure whether I have done it completely right. The directory handling isn't easy to understand there. See the TODOs. 

For now I won't continue with that test script. Other parts seem to be more important.

framps commented:
That's what I expected: You had to update the directory name only and not to change anything in the test logic. I agree it's complicated code. But the test now finishes again successfully and that's important because you updated the SR code and we have to make sure the SR code still works correctly :wink:

framps commented on test/raspiBackup7412Test.sh:
```diff
@@ -145,19 +167,22 @@ function createMassBackups() { # startdate count #per_day type dont_delete_flag
 
 function testMassBackups() { # count type
 
+	#TODO: a) is this function used anywhere?
```
No. testSpecificBackups at the end checks for the normal backups. So this function is a leftover :-(

framps commented on test/raspiBackup7412Test.sh:
```diff
@@ -179,33 +204,45 @@ function testSpecificBackups() { # lineNo stringlist_of_dates type numberOfstati
 	echo "Testing for type $type and static $((2*$static)) ..."
 
-	local f=$(ls $DIR/$(hostname)/ | grep $type | grep -v "_" | wc -l)
+	#TODO: do we need to filter HOSTNAME_OSR?
```
No. This is just a check for the # of directories and it doesn't matter whether there is OSR used or not.

framps commented on test/raspiBackup7412Test.sh:
```diff
@@ -179,33 +204,45 @@ function testSpecificBackups() { # lineNo stringlist_of_dates type numberOfstati
 
-	local f=$(ls $DIR/$(hostname)/ | grep $type | grep -v "_" | wc -l)
+	#TODO: do we need to filter HOSTNAME_OSR?
+	#      or/and do we need to check for old-named backups too?
```
Good catch :+1: . I have to check.

rpi-simonz committed:
framps changes (sent by personal message):
Get getOSRelease by sourcing raspiBackup.sh.
Add test to check for NOT deleting old-named backups.
And some smaller changes.

rpi-simonz committed:
Only show info about old-named backups for a number of times (5)

framps commented on raspiBackup.sh:
```diff
@@ -6301,19 +6310,58 @@ function reportOldBackups() {
+			echo "$DEFAULT_REPORT_COUNTER" > "$report_counter_file"
+			return
+		fi
+		rf=( $(<$report_counter_file) )
```
What's the reason you use `rf=( $(<$report_counter_file) )` instead of just `rf= $(<$report_counter_file)` to read the file contents? I see no reason to use an array here.

rpi-simonz replied:
There is no reason... 
It was a copy&paste from updateRestoreReminder()  ;-)
I'll change it.

framps commented on raspiBackup.sh:
```diff
@@ -6301,19 +6310,58 @@ function reportOldBackups() {
+			if (( $rfn > 0 )) ; then
+				writeToConsole $MSG_LEVEL_MINIMAL $MSG_OLD_NAME_BACKUPS_COUNTER_INFO "${rfn}"
+			else
+				writeToConsole $MSG_LEVEL_MINIMAL $MSG_OLD_NAME_BACKUPS_COUNTER_INFO_BYE
```
It's kindly to say good bye :smile: . But I would just stop the report without any message. That's also done by the restore test reminder and then raspiBackup reminders work consistently.

rpi-simonz replied:
The "good bye" will be removed.

framps replied:
> But the message counter display should/could remain. OK?

I now have time to continue to work on and accept your PR. What do you mean with "message counter display"

rpi-simonz replied:
> What do you mean with "message counter display"?

The message (and related ones):
```
MSG_BACKUP_NAMING_CHANGE=403
MSG_EN[$MSG_BACKUP_NAMING_CHANGE]="RBK0403W: With raspiBackup version %s the naming of the backup directories changed!"
```
will be shown only a couple of times (DEFAULT_REPORT_COUNTER = 5).

And there is a hint about this via message:

`MSG_EN[$MSG_OLD_NAME_BACKUPS_COUNTER_INFO]="RBK0408W: Note: This message will be shown again %s times."`

framps replied:
Thx. Got it. Yes, this message should be kept.

rpi-simonz committed:
Remove the good-bye message again. Remove unneeded array.

framps commented:
Code now rebased into development branch
